### PR TITLE
add: v2

### PR DESCRIPTION
**Pull request by sirpy**
# Description

- using "effectiveStake" for calculations ie stake after donation
- added modifier to require updating global G$ yield per token before calling any update method
- using precision 1e27
- refactored names and added return values

### CODE REVIEW

File: stakingModel/contracts/InterestDistribution.sol [hunk 1]
```diff
@@ -160,7 +162,7 @@ library InterestDistribution {
         Staker storage stakerData = _interestData.stakers[_staker];
         uint256 _withdrawnToDate = stakerData.withdrawnToDate;
 
-        uint256 intermediateInterest = stakerData.totalEffectiveStake.mul(_interestData.globalGDYieldPerToken);
+        uint256 intermediateInterest = stakerData.totalEffectiveStake.mul(_interestData.globalGDYieldPerToken).div(DECIMAL1e18);
```
Comment (sirpy) [inline stakingModel/contracts/InterestDistribution.sol]: why are you dividing by 1e18 all over?
always add comments explaining non trivial calculations

File: stakingModel/test/BasicCalculation.test.js [hunk 2]
```diff
@@ -12,144 +12,127 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
       * Donation% = 30%
-      * TotalStaked = 20 x 1e18
-      * output = 3
+      * GDEarnedInterest = 100 x 1e18
+      * GDEarnedInterestEarned = 100 x 1e18
       */
 
       await interestDistribution
-        .stakeCalculation(user1, web3.utils.toWei("10", "ether"), 30, web3.utils.toWei("1.02", "ether"), web3.utils.toWei("30", "ether"))
+        .stake(user1, web3.utils.toWei("10", "ether"), 30, web3.utils.toWei("100", "ether"), web3.utils.toWei("100", "ether"))
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js]: use numbers that make sense
G$ is in 2 decimals points. so G$ earned interest will not be 100 ether. the earned interest in token should also have logical numbers relative to G$ value

**Review by sirpy — COMMENTED** [review]
@PremSOMISH

File: stakingModel/contracts/InterestDistribution.sol (hunk 1 above)
Comment (PremSOMISH) [reply stakingModel/contracts/InterestDistribution.sol]: Made changes and added comments.

File: stakingModel/test/BasicCalculation.test.js (hunk 2 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: Done

File: stakingModel/test/BasicCalculation.test.js [hunk 3]
```diff
@@ -29,12 +29,12 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
       * Donation% = 30%
-      * GDEarnedInterest = 100 x 1e18
-      * GDEarnedInterestEarned = 100 x 1e18
+      * GDEarnedInterest = 100 x 1e2
+      * GDEarnedInterestEarned = 100 x 1e2
       */
 
       await interestDistribution
-        .stake(user1, web3.utils.toWei("10", "ether"), 30, web3.utils.toWei("100", "ether"), web3.utils.toWei("100", "ether"))
+        .stake(user1, web3.utils.toWei("10", "ether"), 30, 100*100, 100*100)
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js]: still not using logical values
there's an exchange rate between the interest earned in the InterestToken and G$
its not logical that InterestToken earned (100*100) == G$ Earned (100*100)

File: stakingModel/test/BasicCalculation.test.js [hunk 4]
```diff
@@ -43,22 +43,22 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
 
     /**
-      * GDEarnedInterest = 60 x 1e18
-      * GDEarnedInterestEarned = 60 x 1e18
+      * GDEarnedInterest = 60 x 1e2
+      * GDEarnedInterestEarned = 60 x 1e2
       */
     await interestDistribution
-        .updateGlobalGDYieldPerToken(web3.utils.toWei("27", "ether"), web3.utils.toWei("27", "ether"))
+        .updateGlobalGDYieldPerToken(27 * 100, 27 * 100)
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js]: same issue

Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: Looks like you are reviewing older commits, please review recent commit and let me know if issue still exists.

File: stakingModel/test/BasicCalculation.test.js (hunk 3 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: Looks like you are reviewing older commits, please review recent commit and let me know if issue still exists.

File: stakingModel/test/BasicCalculation.test.js [hunk 5]
```diff
@@ -12,144 +12,131 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
-      // 1st stake will not affect AvgYieldRatePerToken and GlobalYieldPerToken because globalTotalStaked and totalStaked both are 0.
+      // 1st stake will not affect GDYieldRate and GlobalYieldPerToken because globalTotalStaked is 0.
       await interestDistribution
-        .stakeCalculation(user1, web3.utils.toWei("20", "ether"), 0, web3.utils.toWei("1", "ether"), web3.utils.toWei("1", "ether"))
+        .stake(user1, web3.utils.toWei("20", "ether"), 0)
         .catch(e => e);
 
+      
+      await interestDistribution
+        .setITokenToTokenRate(web3.utils.toWei("0.4", "ether"))
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js:27]: comment what does this do

File: stakingModel/test/BasicCalculation.test.js [hunk 6]
```diff
@@ -12,144 +12,131 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
-      * Donation = 0%
-      * IntrestTokenRate = 1.03
-      * LastTokenRate = 1.02
-      * IntrestTokenHoldings = 60 x 1e18
+      * GDEarnedInterest = 270 x 1e2
+      * GDEarnedInterestEarned = 270 x 1e2
       */
     await interestDistribution
-        .stakeCalculation(user1, web3.utils.toWei("6", "ether"), 0, web3.utils.toWei("1.03", "ether"), web3.utils.toWei("60", "ether"))
+        .updateGlobalGDYieldPerToken(270 * 100, 270 * 100)
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js]: not logical that they are same number

File: stakingModel/test/BasicCalculation.test.js [hunk 7]
```diff
@@ -12,144 +12,131 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
-      * IntrestTokenRate = 1.03
-      * LastTokenRate = 1.03
-      * IntrestTokenHoldings = 60 x 1e18
+      * Effective stake
       */
-    // Staking so we can manipulate AvgYieldRatePerDAI
+    // Manipulate GDYieldRate to create scenario.
     await interestDistribution
-        .stakeCalculation(user1, web3.utils.toWei("7.2", "ether"), 0, web3.utils.toWei("1.03", "ether"), web3.utils.toWei("60", "ether"))
+        .updateGDYieldRate(user1, web3.utils.toWei("3", "ether"))
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js]: GDYieldRate needs to be renamed
maybe to "stakeBuyinRate"

File: stakingModel/test/BasicCalculation.test.js (hunk 6 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: Isn't it possible if all the stakers are donating 0%?

File: stakingModel/test/BasicCalculation.test.js (hunk 5 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js:27]: done

File: stakingModel/test/BasicCalculation.test.js (hunk 7 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: renamed

File: stakingModel/test/BasicCalculation.test.js (hunk 6 above)
Comment (sirpy) [reply stakingModel/test/BasicCalculation.test.js]: check what are the inputs to updateGlobalGDYieldPerToken, your question is irrelevant

File: stakingModel/test/BasicCalculation.test.js (hunk 5 above)
Comment (sirpy) [reply stakingModel/test/BasicCalculation.test.js:27]: the comment still doesnt explain, the comment should give more information than the method name
setITokenToTokenRate

1. what rate?
2. what do ToToken mean
3. why are you updating the rate here

comments should be descriptive enough so a new developer understand what  it does without digging into the code

File: stakingModel/test/BasicCalculation.test.js (hunk 6 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: Yeah, i know the inputs to updateGlobalGDYieldPerToken() are "Interest earned in G$ after donation" & "Interest earned in G$ before donation". Total donated is 0 till now so I don't understand why both values will not be same.Please elaborate.

File: stakingModel/test/BasicCalculation.test.js (hunk 5 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js:27]: Added comments

File: stakingModel/test/BasicCalculation.test.js (hunk 6 above)
Comment (sirpy) [reply stakingModel/test/BasicCalculation.test.js]: * @param _blockGDInterest           Interest earned in G$ in  exchange for _blockInterestTokenEarned (after donations)
     * @param _blockInterestTokenEarned  Interest token earned (before donations)

one is in GD the other is in InterestToken

Comment (sirpy) [reply stakingModel/test/BasicCalculation.test.js]: interest is earned in interetToken, which is sent to reserve, GD is returned after deducting donations that stay in the reserve

Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: ohh, i thought that both are in GoodDollar just that one is total interest generated and other is interest after removing donation amount. Will fix that and let you know. @sirpy

Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js]: Just clarify one thing..... now _blockGDInterest says (after donations), now if 100 GD is newly minted for interest and from that 100GD, 40 is suppose to go to ubi. now what will be its value? 100 or 60?

File: stakingModel/test/BasicCalculation.test.js [hunk 8]
```diff
@@ -12,144 +12,133 @@ contract("InterestDistribution - Basic calculations", ([user1]) => {
...
 
 
     /**
-      * Stake = 5 x 10^18
-      * Donation = 0%
-      * IntrestTokenRate = 1.03
-      * LastTokenRate = 1.02
-      * IntrestTokenHoldings = 60 x 1e18
+      * GDEarnedInterest = 270 x 1e2
+      * interestTokenEarnedToDate = 1.35 x 1e18
```
Comment (sirpy) [inline stakingModel/test/BasicCalculation.test.js:65]: its not interest token earned to date,
its the interest token earned in the current block just like gdearnedinterest.
still i'm not sure the numbers make sense. 
What is  the simulated exchange rate between the interesttoken and G$? does 1.35(X) is  approximately worth 270GD?

File: stakingModel/contracts/mocks/InterestDistributionMock.sol [hunk 9]
```diff
@@ -45,22 +46,24 @@ contract InterestDistributionMock {
       updatecontractBalance(_amount, false);
       uint requiredCDAIBal = (interestData.globalTotalStaked.sub(_amount)).mul(DECIMAL1e18).div(iTokenToTokenRate);
       uint newGDMinted = 0;
-      uint newGDMintedBeforeDonation = 0;
+      uint excessIToken = 0;
       if(iTokenBalance > requiredCDAIBal) {
-        (newGDMinted, newGDMintedBeforeDonation) = mintGoodDollar(iTokenBalance.sub(requiredCDAIBal));
+        excessIToken = iTokenBalance.sub(requiredCDAIBal);
```
Comment (sirpy) [inline stakingModel/contracts/mocks/InterestDistributionMock.sol]: rename to earnedInterest

File: stakingModel/contracts/InterestDistribution.sol [hunk 10]
```diff
@@ -29,6 +29,7 @@ library InterestDistribution {
         uint256 lastStake;
         uint256 withdrawnToDate;
         uint256 stakeBuyinRate;  // Precision points = 27 + 2 (G$ decimal) = 29
+        uint256 interestClaimedBeforeLastUnstake;
```
Comment (sirpy) [inline stakingModel/contracts/InterestDistribution.sol]: i dont understand why we need this

File: stakingModel/test/BasicCalculation.test.js (hunk 8 above)
Comment (PremSOMISH) [reply stakingModel/test/BasicCalculation.test.js:65]: considering 1cDai = 200 G$

File: stakingModel/contracts/InterestDistribution.sol (hunk 10 above)
Comment (PremSOMISH) [reply stakingModel/contracts/InterestDistribution.sol]: While unstake, we are resetting withdrawnToDate so to preserve staker's all time claimed interest we need one variable. Let me know if we don't need to preserve staker's all time claimed interest, will  remove it.

Comment (sirpy) [reply stakingModel/contracts/InterestDistribution.sol]: we dont need it.
you can merge PRs after you remove it

File: stakingModel/contracts/mocks/InterestDistributionMock.sol (hunk 9 above)
Comment (PremSOMISH) [reply stakingModel/contracts/mocks/InterestDistributionMock.sol]: done

File: stakingModel/contracts/InterestDistribution.sol (hunk 10 above)
Comment (PremSOMISH) [reply stakingModel/contracts/InterestDistribution.sol]: removed

**Pull request merged**